### PR TITLE
feat: Draft & Publish V5

### PR DESCRIPTION
**Pull request by Marc-Roig**
## ❓  What does it do?

There are many rough edges to polish, but this an iteration of handling Documents and D&P in the content manager.

The most important changes:
- Refactored Collection-Types contracts api to return document metadata (available locales and publication status of a document)
- Refactored entity-manager to work with the document service (we will rename it to document-manager)
- Refactored collection-types controller to work with documents.

## What still doesn't work

- Bulk actions (publish, unpublish, delete)
- Relations ( we still need to work some parts on the document service)\
- Publish documents with unique fields (fields are not unique across every entry but across the entire document at the moment).
- Publish/Unpublish many locales at once
- I overlooked permissions, and sanitization.
- Validations on clone / publish are not working correctly. (required, min, maxlength, ...)
- Review workflows
- Non localized attributes

## 🧪  Test Endpoints

<details>
<summary><h3>Find Tag</h3></summary>

`GET  /content-manager/collection-types/api::tag.tag/<DOC_ID>`

```ts
type QueryParams = {
	locale: string
	status: 'draft' | 'published'
}
```

</details>

<details>
<summary><h3>Find Many Tags</h3></summary>

`GET  /content-manager/collection-types/api::tag.tag`

```ts
type QueryParams = {
	locale: string
	status: 'draft' | 'published'
}
```

</details>

<details>
<summary><h3>Create a draft tag</h3></summary>

`POST /content-manager/collection-types/api::tag.tag`

```ts
type Body = {
	locale: string
} & Document
```

</details>

<details>
<summary><h3>Update a draft tag</h3></summary>

`PUT /content-manager/collection-types/api::tag.tag/<DOC_ID>`

```ts
type Body = {
	locale: string // If the provided locale does not exist, but the document id does, it will create the locale
} & Document
```

</details>

<details>
<summary><h3>Publish a draft tag</h3></summary>

`POST /content-manager/collection-types/api::tag.tag/<DOC_ID>/actions/publish`

⚠️  Publishing many locales at the same time is still not supported
```ts
type Body = {
	locale: string
} 
```

</details>

<details>
<summary><h3>Unpublish a tag</h3></summary>

`POST /content-manager/collection-types/api::tag.tag/<DOC_ID>/actions/unpbulish`

⚠️  Unpublishing many locales at the same time is still not supported

```ts
type Body = {
	locale: string
} 
```

</details>

<details>
<summary><h3>Delete a tag</h3></summary>

`DELETE /content-manager/collection-types/api::tag.tag/<DOC_ID>`

```ts
type Body = {
	locale?: string // if not provided, it will delete the entire document
} 
```

</details>

### CODE REVIEW

**Comment by Marc-Roig** [comment]
I will mark this as ready to be reviewed, but next week I will be working on the unit and api tests.

There are many things to polish here, but now is at a state where it's usable for the UI to start working on D&P. 
@markkaylor if you are available next week we could sync the document service work and what we have done in the content manager 🚀

**Comment by Marc-Roig** [comment]
I also plan to refactor a bit the document metadata service, I don't really like it atm. Feel free to propose ideas 🚀

**Review by innerdvations — COMMENTED** [review]
Code LGTM but I didn't test that it works

File: packages/core/content-manager/server/src/services/document-metadata.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
+import type { LoadedStrapi as Strapi, Common } from '@strapi/types';
+import type { DocumentMetadata } from '../../../shared/contracts/collection-types';
+
+export interface Document {
+  id: string;
+  locale: string;
+  status: string;
+  publishedAt: string;
+}
```
[the remark this answers is not in the document]
Comment (Marc-Roig) [reply packages/core/content-manager/server/src/services/document-metadata.ts:9]: Absolutely 🙌

**Comment by Marc-Roig** [comment]
@jhoward1994 You will see this error on every content type without i18n enabled, forgot to mention that in the description!

File: packages/core/content-manager/server/src/services/entity-manager.ts [hunk 2]
```diff
@@ -55,7 +54,15 @@ export default ({ strapi }: { strapi: Strapi }) => ({
    * @param {Object} entity
    * @returns
    */
-  mapEntity<T = unknown>(entity: T): T {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  mapEntity<T = any>(entity: any, uid?: Common.UID.ContentType): T {
```
Comment (innerdvations) [inline packages/core/content-manager/server/src/services/entity-manager.ts:58]: Entity -> document here

File: packages/core/content-manager/server/src/services/entity-manager.ts [hunk 3]
```diff
@@ -180,31 +241,22 @@ export default ({ strapi }: { strapi: Strapi }) => ({
...
-
-    await emitEvent(uid, ENTRY_PUBLISH, updatedEntity!);
+    const { versions: publishedDocuments } = await strapi
+      .documents(uid)
+      // @ts-expect-error - Change entity to document
+      .publish(document.id, params);
 
-    // @ts-expect-error mapEntity can be extended
-    const mappedEntity = await this.mapEntity(updatedEntity, uid);
+    // TODO: What if we publish many versions at once?
```
Comment (innerdvations) [inline packages/core/content-manager/server/src/services/entity-manager.ts:257]: IMO we either return the default locale unless they passed one in, or nothing at all.

**Comment by jhoward1994** [comment]
For creation I need to specify the locale as query params and not in the request body as you mention e.g.


`http://localhost:1337/content-manager/collection-types/api::tag.tag?plugins[i18n][locale]=fr`

**Comment by Marc-Roig** [comment]
@jhoward1994 Sorry! we still did not merge this PR https://github.com/strapi/strapi/pull/18975 which actually fixes that

**Comment by Marc-Roig** [comment]
@jhoward1994 @innerdvations I have a list of todos based on your comments ! Thank you for the reviews

I will proceed with the tests and create other pr's that add your feedback ( if it was not a bug you found or something broken)

File: packages/core/content-manager/shared/contracts/collection-types-v5.ts [hunk 4]
```diff
@@ -0,0 +1,322 @@
+// TODO: Use this contract when we make the switch to the new V5 API
+import { errors } from '@strapi/utils';
+import { Schema, Common, Documents } from '@strapi/types';
```
Comment (Marc-Roig) [inline packages/core/content-manager/shared/contracts/collection-types-v5.ts:3]: @jhoward1994 @joshuaellis I decided to create two new parallel -v5 file contracts for content manager. So when we decide to switch to the new api, we have this ready :) 

I left the original contracts as they are, as it is what we are actually returning at the moment.

Comment (joshuaellis) [reply packages/core/content-manager/shared/contracts/collection-types-v5.ts:3]: Good idea!!

**Pull request merged**